src/khiva/matrix.cpp

```diff
@@ -746,12 +499,244 @@ void findBestN(af::array profile, af::array index, long m, long n, af::array &di
+
+    af::array sum_t = cumulative_sum_t(af::seq(m, na), af::span) - cumulative_sum_t(af::seq(0, na - m), af::span);
+    // Cumulative sum of the element-wise square of each subsequence of all the time series contained in t
+    af::array sum_t2 = cumulative_sum_t2(af::seq(m, na), af::span) - cumulative_sum_t2(af::seq(0, na - m), af::span);
+
+    // Mean of each subsequence of all the time series
+    mean = sum_t / m;
+    // Mean of the element-wise square of each subsequence of t
+    af::array mean_t2 = sum_t2 / m;
+    // Square of the mean
+    af::array mean_t_p2 = af::pow(mean, 2);
+    // Variance
+    af::array sigma_t2 = mean_t2 - mean_t_p2;
+    // Standard deviation
+    stdev = af::sqrt(sigma_t2);
+}
+
+void calculateDistanceProfile(af::array qt, af::array a, af::array sum_q, af::array sum_q2, af::array mean_t,
+                              af::array sigma_t, af::array mask, af::array &distance, af::array &index) {
+    long batchSize = static_cast<long>(qt.dims(3));
+    long tsLength = static_cast<long>(qt.dims(0));
+    long nTimeSeries = static_cast<long>(qt.dims(1));
+
+    // Tiling the input data to match the batch size, the time series length and the number of time series
+    af::array a_tiled = af::tile(a, 1, 1, 1, static_cast<unsigned int>(batchSize));
+    af::array sum_q_tiled =
+        af::tile(sum_q, static_cast<unsigned int>(tsLength), static_cast<unsigned int>(nTimeSeries));
+    af::array sum_q2_tiled =
+        af::tile(sum_q2, static_cast<unsigned int>(tsLength), static_cast<unsigned int>(nTimeSeries));
+    af::array mean_t_tiled = af::tile(mean_t, 1, 1, 1, static_cast<unsigned int>(batchSize));
+    af::array sigma_t_tiled = af::tile(sigma_t, 1, 1, 1, static_cast<unsigned int>(batchSize));
+
+    // Required to avoid a division by zero when the standard deviation is zero
+    double eps = (sigma_t_tiled.type() == 0) ? EPSILON * 1e4 : EPSILON;
+    af::array lessThanEpsilon = eps >= sigma_t_tiled;
+    sigma_t_tiled = lessThanEpsilon * lessThanEpsilon.as(sigma_t_tiled.type()) + !lessThanEpsilon * sigma_t_tiled;
+    // Computing the distance
+    af::array dist = a_tiled + (-2 * (qt - sum_q_tiled * mean_t_tiled) / sigma_t_tiled) + sum_q2_tiled;
+    dist = af::sqrt(af::abs(dist));
+
+    // The 1st dimension reflects the number of subsequences of the reference time series.
+    // The 2nd dimension reflects the number of query time series.
+    // The 3rd dimension reflects the number of reference time series
+    // The 4th dimension reflects the batch size of query subsequences from the query time series
+    // Reordering to match the mask band matrix dimensions
+    dist = af::reorder(dist, 3, 0, 1, 2);
+
+    // Increasing the distance using the mask band matrix to filter trivial matches
+    dist += 1 / EPSILON * mask.as(qt.type());
+
+    // The 1st dimension reflects the number of subsequences of the reference time series.
+    // The 2nd dimension reflects the number of reference time series.
```

Comment:
Please be consistent with the comments, if you are going to add a point at the end of the comment, add it to every one, if not, write all the comments without it.